docs/apis/core/customfields/index.md

```diff
@@ -0,0 +1,168 @@
+
+Handler has protected constructor, to get a handler call create() method. Some areas may choose to return a singleton here:
+
+```php
+$handler = HANDLERCLASS::create($itemid);
+```
+
+Configuration page contents will be:
+
+```php
+$output = $PAGE->get_renderer('core_customfield');
+$outputpage = new \core_customfield\output\management($handler);
+echo $output->render($outputpage);
+```
+
+Handler must implement all abstract methods (calculate configuration or instance context, check permissions to configure, view or edit) and also may choose to overwrite:
+
+```php
+handler::uses_categories()
+handler::generate_category_name()
+handler::config_form_definition() // For example, the course_handler adds "locked" and "visibility" settings that control who can edit or view the particular field.
+handler::setup_edit_page() // Sets page context/url/breadcrumb for the customfield/edit.php page, in some cases it must be overridden.
+```
+
+### Add custom fields to the instance edit form
+
+Custom fields are added to the **instances**. For example, course custom fields are added to the courses, so courseid is the instanceid. In the example of [mod_surveybuilder](https://github.com/marinaglancy/moodle-mod_surveybuilder) we use $USER->id as the instanceid (which means that in this example one user can fill the survey in one module only once). In each case of using custom fields there should be a clear concept of an **instance**. Instanceid is required to save the data but it may be empty when we render the instance edit form (for example, the course is not yet created).
+
+Developer must add custom fields callbacks to the instance edit form. If the instance is "made up" (like in mod_surveybuilder), a new form has to be created with 'id' field in it that will refer to the instanceid.
+
+The following callbacks should be used in form definition, definition_after_data, validation and after form submission:
+
+```php
+$handler->instance_form_definition($mform)
+$handler->instance_form_before_set_data()
+$handler->instance_form_definition_after_data()
+$handler->instance_form_validation()
+$handler->instance_form_save($data)
+```
+
+The `instance_form_save()` method must be called after the form was saved as the `$data` parameter must have the `id` attribute.
+
+On deletion of an instance or on deletion of the whole item call:
+
+```php
+$handler->delete_instance()
+$handler->delete_all()
+```
+
+### Retrieving instances custom fields
+
+How custom fields are used depends entirely on the situation. The following handler methods will help you to retrieve custom fields values for the given instance(s):
+
+```php
```

Comment:
```suggestion
```php title="Handler methods to retrieve custom fields values for the given instance(s)"
```